New Upstream: SX 1.2

edwintorok opened the pull request:
Changes since https://github.com/gcsideal/sx-debian/pull/2:
- SX 1.2
- `libsx` renamed upstream to `libsxclient` to avoid conflict with `libsx-dev`/`libsx0` package.
- enable PIE and BINDNOW

Commit:
Improve sxls efficiency (bb#415)

This improvement let database to take advantage of indexing and reduces
redundant queries. More text processing is done inside sqlite engine
than it was done before.

Commit:
fix make check when --disable-sxhttpd is used

enable proxy and scgi modules to allow to unconditionally set temp path

Commit:
Store tmpfile revision in job data for file delete jobs (bb#1005)

Commit:
Add node status query (bb#1112)

Commit:
add --desc to userclone

Commit:
client-test.c: use proper size for string (bb #1135)

it needs space for the \0, better let the compiler calculate the correct size
than hardcoding a value off-by-one.

Commit:
make check: use shorter path for socket name (bb #1134)

The maximum length of a socket path is small (UNIX_PATH_MAX=108 on Linux),
so use a relative path to avoid 'make check' failing when run from different
build paths.

Also print the fcgi log always, even when starting the fcgi fails.

Silence mkdir error.

Commit:
fix more path limits (bb #1134)

minimum value for PATH_MAX is _XOPEN_PATH_MAX=1024, so use that
shorten path in nginx too (use relative path)

Commit:
fix build on BSD (bb #1134)

Commit:
backport fix for bb #1134

the build sometimes failed when the build path was too long,
use relative path to avoid this.

Commit:
Allow to bump revisions limit for volume (bb#1127)

Revisions decrease schedules background job on all volnodes in cluster.
The job drops all too old revisions of files stored in volume.

Add also some hashfs comsmetics (mostly comments changes).

Commit:
flag to detect copying a file to itself (bb #1048)

always enabled for local files.

Commit:
fix typo

gcsideal committed:
Merge pull request #1 from edwintorok/debian

Needs tuning, but basically correct.

Commit:
Fix sxreport-server segfault on bare nodes

Commit:
identify more race conditions on cluster join

temporarily locked isn't the only possible message

Commit:
do not override job_new error reason

Commit:
enable uwsgi

Commit:
set SXHTTPD too

Commit:
fix error path leak

gcsideal committed:
Standards-Version is 3.9.6

gcsideal committed:
Add up-to-date changelog with ITP bug number

Commit:
add check for minimum node size (bb#919)

Commit:
fix undopriv (bb#1086)

Commit:
Add ignore_errors flag to sxrm (bb#1006)

Commit:
Avoid falling into infinite recursive loop when client fails to upload hashes (bb#878)

Commit:
Shuffle host list before scheduling delete job (bb#860)

Commit:
Introduce timeouts for block transfers (bb#981)

Commit:
fix libfcgi includes when building out of tree

Commit:
Reorder locking databases in sx_hashfs_check() (bb#1128)

It reduces possibility to fall into a race while deleting files.

Commit:
Improve file delete efficiency with better indexing.

This fixes efficiency loss introduced by 380d2c706f52882653708db25b6bb238d935a5fc.
Create index in data field in jobs table and split qt_tmpbyrev query to
use indices again.

Commit:
GC references instead of counters

turn the operations table into a table of revision -> block references
drop the 'used' table that contains the counters
add a table that track the operations on revisions

This is required to support volumes with files that have different replica
counts due to cluster temporarily modified by marking faulty nodes.
It also means that it'll be safe to GC hashes that are shared between
files that were part of a dead replica 1 node, and replica >1 files.

We can GC a block when it has no references.
We can GC a reference (revision) when the sum of its operations is 0.
Keep operations with negative counters (deletes) to avoid race conditions during rebalance.
Rules for GC of reservations is as before: if its TTL expires, or if there was no upload activity on it for <grace_period>.

Make reservation, flush and delete IDs match:
 a reservation is like a +1 except it also has an entry into the reservation table
 the id (hashes) of these operations must match:
  - reserve_id: based on volume/filename so that successfully uploading a file deletes all reservations for it (as before)
  - revision_id: hash of the file's revision and more info to make it unique across the cluster

TODO: upgrade job from 1.0
TODO: allow deleting files even if it has only negative revisions (repeated/replayed deletes)
TODO: optimize file delete: unbump only revision not all hashes
TODO: optimize file flush: bump only revision not all hashes, just do a presence check on the hashes

Commit:
Add --info option to sxinit (bb#440)

Commit:
Compare crc32 of each block while uploading file

This reduces possibility of uploading files which has changed while
begin uploaded and therefore storing spurious data on cluster.

Commit:
Implement advisory locking (bb#569)

Commit:
check for sys/statvfs.h (bb#1160)

Commit:
sxsetup: use ip instead of ifconfig (bb#851)

Commit:
GC references: Update replication, rebalance and repair API

updated s2s API to send the revision hash for each hash instead of counters.
We send: block hash, blocksize, +-, revision hash, replica#

TODO: investigate delete during rebalance, add more age < restriction as required

Commit:
move error check before use of uninit var

Commit:
send proper revision_id for replicas

Commit:
correct suggestions to be compatible with older man tools (bb#1163)

Commit:
use proper column for reservation expiration

Commit:
delete all unused revisions when reservation is deleted

Commit:
fix repair: don't overwrite the loop variable

Commit:
Add owner entry to sxls -l sx://<cluster>

Commit:
delete just revision on each node instead of unbumping blocks

Commit:
WiP: file flush: use new GC APIs

Commit:
sxadm --info: during rebalance print removed nodes too (bb#1148)

Commit:
Do not fail --unlock when cluster is rebalancing

Commit:
sxsetup: lock cluster before calling --mod

Commit:
Reduce sleeping on file flush operation. (bb#331)

Commit:
upgrade-from-10: move DB schema modification to a separate function

Commit:
perform integrity check before schema upgrade

Commit:
integrity check after upgrade too

Commit:
Implement cluster read-only/read-write modes switching (bb#542)

Commit:
Improve some hashfs logging messages. (bb#1007)

Commit:
Add degraded mode support

Commit:
local file blocks upgrade 1.0->1.1

TODO:
 - node upgrade command (or automatic)
 - insert proper revision_id hash when creating files
 - disable GC until upgrade job completes
 - better progress in sxadm job poll, automatically chain upgrade job to previous one
 - create heal entries for remote file blocks too
 - process remote heal entries in gc process
 (which should be renamed to heal&gc)

Commit:
Job polling cosmetics

Commit:
Allow to perform sxadm node --check when cluster is in read-only mode, otherwise reject it (bb#1155)

Commit:
various cleanups required on gcpartial

allow job_new_2pc to create child jobs
change type of spec->nodes to take hashfs

Commit:
various cleanups required by revision delete

tmp_getinfo doesn't need op_expires_at parameter
null checks
blocksize for file_totmp

Commit:
add dummy jobs

Commit:
add dummy job placeholder for setfaulty

Commit:
sxsetup: create a remote copy of sxsetup.conf and use it to obtain SSL config for new nodes

Commit:
check for NULL in DEBUGHASH caller

Commit:
fix compiler warnings

Commit:
fix compiler warnings

Commit:
When using system nginx disable gzip so that ETag works

Commit:
sxls --etag support

store etag in ~/.sx/volumes/<vol>/etag/<etagfile>

you can track etags separately for different purposes (sync uploaded file,
 deleted files, etc.) with different etag files.

Commit:
add sxc_lib_(init|shutdown) and sxc_client_(init|shutdown) for mt apps

Commit:
Bump expiration time for expired abort/undo jobs (bb#886)

Commit:
Add request timeouts server side. (bb#1157)

Commit:
Check volume owner length

Commit:
Rollback transaction on error condition

Commit:
remote volume heal WiP

Commit:
drop file_totmp and fix filedelete_{undo,abort}

they must call undo/abort on the revision unbump job

Commit:
upgrade for remote file blocks

node upgrade command
enable foreign keys for events.db
perform integrity check and only upgrade after that
allow to run local upgrade job directly with hidden flag
events cleanup (bb #1201)
fix warnings
cleanup and error handling
automatically insert upgrade job when node --upgrade is run
add NOT NULL constraint to new column
add upgrade step that can depend on all DBs.

temporarily disabled

Commit:
Implement cluster rename (bb#1138)

Commit:
Implement sxacl whoami --role (bb#1213)

edwintorok committed:
Fix gcc 5.0 warning: use __func__ instead of __FUNCTION__

Commit:
Implement token generation based on username, password and cluster UUID (bb#1215)

Commit:
manual: add support for conversion with tex4ht, thanks to Nicola Torracca

Commit:
Support configuration links

Commit:
Allow hashfs check to be called when node is down

Commit:
fix leak on large file listing

Commit:
split tmpfileid_or_rev

Commit:
Add --pass-file option for sxint and sxacl useradd/usernewkey

Cosmetics: print configuration link in sxinit --info

Commit:
warning fixes

Commit:
Degraded mode fixes

Commit:
Report degraded status to sxadm

Commit:
Decrease number of bcrypt iterations to 2^12 for user token generation

Commit:
3rdparty: upgrade to sqlite 3.8.9

Commit:
Merge remote-tracking branch 'origin/setfaultysqash'

Commit:
use sxinit --key instead of sxinit in start script

Commit:
forbid control characters and path separators in user and volume names

Commit:
fix logrotate script (bb#934)

Commit:
Do not use hashfs volumes listing api for sxadm node --extract (bb#1240)

Commit:
re-enable upgrade, and allow to terminate cleanly

Commit:
Upgrade filedelete jobs

Commit:
Merge remote-tracking branch 'origin/faultymerged'

Commit:
fix warning on freebsd

Commit:
Check status of all nodes despite communication errors (bb#1255)

Commit:
replica cannot be zero

Commit:
set reserve replica

Commit:
Check hdist changes status before locking distribution

Commit:
Check if UID was returned by server (bb#1254)

Commit:
Use sxi_strlcpy instead of memcpy in filerev_from_jobdata_tmpfileid

Commit:
don't create a new revision if the file hasn't changed

Commit:
Support sxauthd in sxinit (bb#1225)

Commit:
use sxi_rand_pseudo_bytes for generating UUIDs

Commit:
Clear error buffer when single_download() succeeds after multi_download() failure (bb#1186)

Commit:
set SQLITE_SCHEMA messages to level debug

Commit:
handle stdout output in sxrev copy

Commit:
sxsetup compatibility with the latest sxinit

Commit:
Send operating mode with .sync query (bb#1269)

Commit:
don't fail when upgrading bare node

Commit:
sxadm cluster --upgrade

also do not allow to start remote upgrade while local upgrade hasn't completed yet

Commit:
add manpages for server tools (bb#1136)

Commit:
3rdparty: upgrade curl to 7.42.1

Commit:
fix iteration

Commit:
ignore ENOENT on delete

Commit:
avoid git -C (bb #1153)

fixes libsx version on systems with old git (CentOS, Debian wheezy, etc.)

Commit:
Do not fail r2rfast while cluster is rebalancing (bb#1268)

Commit:
cosmetics for 1.1 (bb#993)

Commit:
iterate on gc_max_batch items at once instead of just 1

Commit:
don't error if wait is 0 and job status is PENDING

and add DEBUG messages

Commit:
close GC trigger descriptors (bb #1189)

Commit:
fix warnings

Commit:
handle output from the latest sxadm (bb#1274)

Commit:
sxadm node --owner should work for all node subcommands

delay user switch only for create, which might need to read a key that is not
accessible for the new user.

Commit:
sxsetup: initial support for --upgrade

Commit:
add upgrade status to sxadm cluster --status, and cluster --upgrade (bb #1279)

Commit:
bump library soname (bb #1250)

Commit:
bump SXF_ABI_VERSION too

Commit:
make sure all volumes are processed when there is more than one volume/metadb

Commit:
handle uuid=NULL

Commit:
more debug code

Commit:
make a copy of sx_hashfs_volume_t

it can change while we're using it because it points to the same member in
sx_hashfs_t

Commit:
don't reset rebalance state, use separate state for upgrade progress

Commit:
sxadm node --upgrade: don't print error about already existing upgrade job

Commit:
ensure node not running (bb #1279)

Commit:
check that all nodes are upgraded

Commit:
sxsetup --upgrade improvements (bb#1279)

Commit:
skip hashes with count=0 on rebalance iteration retry (instead of failing)

Commit:
Get tempfile meta from database once

Commit:
close dbf only if it was successfully open

Commit:
fix newline in cluster --status

Commit:
hide heal status from user

Commit:
downgrade 'upgrade aborted' severity

this was just a status query that didn't take any action

Commit:
avoid possible null deref (bb #1281)

Commit:
cosmetics

Commit:
fix error message

Commit:
fix error path debug msg

Commit:
drop sql query plan tests (they are implemented in hashfs.c already)

Commit:
log delete undo errors (bb #1256)

Commit:
add more checks in jobpoll

Commit:
3rdparty: upgrade to sqlite 3.8.10

Commit:
user larger buffer in randgen

Commit:
fix etag_out leak

Commit:
properly report invalid argument (bb#1304)

Commit:
update documentation for 1.1
incorporate sxdump
add man pages for server tools

Commit:
use local sxhttpd even if prefix is /usr

otherwise it tries to use /usr/sbin/sxhttpd which doesn't exist yet

Commit:
allow retry on data upload fail

Commit:
Update for SX 1.1

drop patches (merged upstream)
libsx has SONAME libsx.so.2 now
fix logrotate script
runtime-depend on python (/usr/sbin/sxdump is written python)
runtime-depend on ca-certificates (needed for SSL cluster to work properly)
update gitignore
sxsetup now provides an --upgrade command
upstream includes manpages for server tools too
update copyright years
do not restart on upgrade (requires manual intervention)

Commit:
fix warning about %p

Commit:
update gitignore

Commit:
fix aes256 build on OS X

Commit:
bump version number to 1.1

and regenerate build sys with automake 1.14.1

Commit:
sxdump: when changing cmdline flag the field has to be renamed too

Commit:
cosmetics

Commit:
aes256: don't ask twice for password in decryption mode when 'nogenkey' is used

Commit:
skip volumes with unknown filters

Commit:
Imported Upstream version 1.1+ds

Commit:
Merge tag 'upstream/1.1+ds' into debian

Upstream version 1.1+ds

Commit:
Implement batched delete (bb#931)

Commit:
sxmv: when filenames differ only by one letter they are still different

edwintorok committed:
tweak TLS ciphers

Commit:
sxsetup: check total memory and show a warning/limit number of children processes (bb#1307)

Commit:
add datetime support in blobs

Commit:
support IPv6 in hdist (bb #1331)

Commit:
enable ipv6 in sxhttpd

Commit:
remove blocks that become unused during rebalance (bb #1326)

when the block was processed the first time it was still in use,
but by the time we retry it became unused, and has to be removed from the retry
list.

Commit:
sxcp: set mtime on all files (bb #1353)

Commit:
user api: return proper http code on errors other than 404

Commit:
Correct unique file id generation

Commit:
miscellaneous fixes and other cosmetics

Commit:
fixup: don't put a transaction inside a transaction dawg

Commit:
Implement quota enforcement per volume owners (bb#1231)

Commit:
use hashfs upgrade facility, some sql rework

Commit:
Fix new volumes privs assignment

Handle situation when volume owner's clones exist in cluster. When
assigning a volume to a clone this commit avoids putting unnecessary
privs table entry.

Commit:
Size enforcement rework

Consists of the following changes
- Allow same-file reupload (bb1342)
- Avoid size evasion via meta+extend
- Properly check for new revision
- Do size computations in SQL

Commit:
Add support for changing user description (bb#1242)

Commit:
Implement custom volume meta handling (bb#1242)

Commit:
Drop unused sizeOnly parameter (bb#1313)

Commit:
client-test: Add user quota and custom volume meta tests

Fix plenty of compiler warnings

Commit:
reuse token on extend (bb#1367)

Commit:
add back the old error message for 'Volume already exists' (bb #1407)

Commit:
Allow users having write access to use mass delete

Commit:
Rework custom volume meta handling

Get rid of sxc_meta_t structure inside hashfs related operations,
fix minort issues.

Commit:
Introduce cluster meta support (bb#1325)

Commit:
Decode plus encoded space in url params

Commit:
use smart request ids (bb#1413)

Commit:
rename libsx to libsxclient (bb#1412)

Commit:
Handle '0 remote volume names listed' on upgrade (bb #1311)

When upgrading SX 1.0 -> 1.1 if there are no more remote volumes
(e.g. all volumes have replica count = node count) then set upgrade status to
finish, instead of waiting 1h for the GC to be triggered.

Commit:
add cleanup code to fcgi-test

Commit:
add support for custom volume meta to filter API

Commit:
aes256: use custom meta to store key fingerprint (bb#1403)

Commit:
Avoid undefined behaviour in signed left-shift

Due to C integer promotion rules data[3] in 'k1 |= data[3] << 24' is first promoted to int,
then shifted left, then converted to unsigned.
Signed left shift with 24 is undefined because it might change the sign bit.

Just explicitly cast data[3] to avoid promoting unsigned char to int.
Fixes warnings shown by -fsanitize=undefined.

Commit:
fix minor warnings and error path leaks (bb#1361)

Commit:
minor fixes

Commit:
use ldconfig -X without $(libdir) (bb#1387)

Commit:
fix make check in some cases (bb#1334)

Commit:
PRIV_MANAGER support (bb #1357)

sync all acls to bare node properly
refactor acl parsing: move to sxacl.c
cosmetics
automatically grant/revoke RW privileges for manager
send api-version: 2 for acl
allow only owner or admin to grant/revoke the manager privilege
fix make check: owner implies manager privilege

Commit:
use sxi_hmac_sha1_cleanup() instead of sxi_md_cleanup()

Commit:
If possible use locally stored authorization token in sxacl usergetkey (bb#1414)

Commit:
check faulty and replaced nodes and warn if they are found out to be alive

Commit:
rebalance progress reporting and iteration sanity checks

Commit:
Fail when sxc_cluster_listvolumes_next is used incorrectly (bb#1405)

Commit:
typo in warn

Commit:
sxvol modify: add new options --reset-custom-meta and --reset-local-config

Commit:
aes256: by default don't ask for a password when creating a volume

The key and its fingerprint will be set during the first access to the volume,
and can be later reset if needed. The administrator can also assign a permanent
key using the 'setkey' option.

Commit:
Add .self query, replacement for ?whoami (bb1366)

Commit:
Allow volume owner to change revisions limit (bb#1428)

Commit:
Print per-file errors only when ignore-errors is set

Commit:
make sxi_file_dup() static

Commit:
Implement mass rename (bb#1317)

Commit:
rerun 1.0 -> 1.1 upgrade job on 1.1 -> 1.2

Mass rename requires that revision_id doesn't depend on filename.
Have to recalculate all revision ids and rebuild the file->block maps
because rebalance doesn't preserve the legacy revision ids.

Commit:
Use revision_id stored in files database

Commit:
Add revision ID self-checking code

Commit:
Use EAGAIN instead of ECOMM

Commit:
bump all API versions

Commit:
drop unused parameter

Commit:
check arguments to CGI_PRINTF

Commit:
fix make check for centos builds

update SXHTTPD sed
(SXHTTPD is /usr/sbin/sxhttpd and overriding sbindir is not enough)

Commit:
compatibility with SX 1.1 on all APIs

update make check

Commit:
update sxacl --help to include 'manager' privilege

Commit:
sxvol modify --owner revokes all privileges from previous owner

Commit:
update sxvol modify manpage

Revert "sxvol modify --owner revokes all privileges from previous owner"

Commit:
aes256: show a specific msg when setting a volume password

Commit:
fix description of --set-mode

Commit:
update docs

Commit:
Make userlist API compatible with 1.1

Commit:
Disable revision ID debugging

Commit:
bump to SX 1.2

Commit:
Imported Upstream version 1.2+ds

Commit:
Merge tag 'upstream/1.2+ds' into debian

Upstream version 1.2+ds

Commit:
Update to new upstream version SX 1.2

libsx renamed upstream to libsxclient to avoid conflict with libsx-dev/libsx0 package.
enable PIE and BINDNOW

gcsideal commented:
Sorry, didn't realized first that you further upgraded SQLite3 to version 3.8.10. Still it would be nice if you consider using version <a href="http://www.sqlite.org/releaselog/3_8_11.html">3.8.11</a>:
- Enhance the page cache so that it can preallocate a block of memory to use for the initial set page cache lines. Set the default preallocation to 100 pages. Yields about a 5% performance increase on common workloads.
- Miscellaneous micro-optimizations result in 22.3% more work for the same number of CPU cycles relative to the previous release. SQLite now runs twice as fast as version 3.8.0 and three times as fast as version 3.3.9.

Also see the sentences under "Important bug fixes". In general it would be wiser if you let the package find system libraries instead of shipping old / vulnerable / performance problematic versions of 3rd party software.

edwintorok commented:
@gcsideal Sqlite3 3.8.11 was just released just today, after we released SX itself (or at least the announcement mail was). Thanks for the heads up we'll definetely upgrade it in master and for the next release.
This shouldn't be a concern for Debian and Fedora because the 3rdparty/ folder is stripped and a configure flag is used to ensure the system library is used always (provided that it meets the version requirement `>= 3.8.4.3`). Sorry for the confusion, my pull request includes our full git revision history, even for things that are excluded by pristine-tar, if you checkout the head of the debian branch there should be no `3rdparty/` folder and no `sqlite3.c`.
The relevant commits for packaging purposes should be: https://github.com/edwintorok/sx-debian/commit/ec3c3a7ae2cecf4639f3ae38dd0b6f44b58ea06c and the older https://github.com/edwintorok/sx-debian/commit/69f981a1c839fffc2425f316ab411ee60277937c, but you are of course welcome to go through our git logs as well.

Commit:
add libjson-perl as a dependency

needed when installing minimal set of packages with apt --no-install-recommends

Commit:
add patch for make check when /proc is not mounted
